src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -655,6 +655,22 @@ SubCmdProgram::execute(const SubCmdOptions& _options) const
     return;
   }
 
+  // We support programming all devices only for a u30 homogeneous setup
+  // --device all is deprecated 
+  for (const auto& dev : deviceCollection) {
+    if(deviceCollection.size() != 1 && xrt_core::device_query<xrt_core::query::board_name>(dev).compare("u30") != 0) {
+      std::cerr << "\nERROR: Multiple device programming is not supported. Please specify a single device using --device option\n\n";
+      std::cout << "List of available devices:" << std::endl;
+      boost::property_tree::ptree available_devices = XBU::get_available_devices(false);
+      for(auto& kd : available_devices) {
+        boost::property_tree::ptree& dev = kd.second;
+        std::cout << boost::format("  [%s] : %s\n") % dev.get<std::string>("bdf") % dev.get<std::string>("vbnv");
+      }
+      std::cout << std::endl;
+        return;
```

Comment:
Going forward, instead of simply returning from a SubCmd if there is an issue, throw the following exception instead:
`
    throw xrt_core::error(std::errc::operation_canceled);
`

Doing so will allow an error value to be returned by the tool.

Note: Currently, only a few SubCmds have been refactored with this new "functionality."   I am planning on completing the remaining ones over the next week.